### PR TITLE
JP-2243: Implement MIRI LRS pathloss correction

### LINKED ISSUE #6277: Throughput correction for MIRI LRS slit observations of point sources
_Issue [JP-2243](https://jira.stsci.edu/browse/JP-2243) was created on JIRA by [Sarah Kendrew](https://jira.stsci.edu/secure/ViewProfile.jspa?name=skendrew):_

The LRS team has become aware that in our current calibration process we are not correcting for the slit loss for point source observation - that is, the wavelength-dependent fraction of flux lost due to the PSF being truncated by the slit mask in the nominal pointing locations. We have determined these loss factors by analysis using WebbPSF and have created a calibration file to perform the correction. This should be implemented as part of extract1d, for SLIT observations of POINT sources only. 

I have assigned this as <ins>high</ins> priority, success of commissioning does not critically depend on this being in the pipeline as we can apply the extra correction manually; however it is an important aspect of the calibration. The CDP file is attached to the ticket and will likely be in our upcoming CDP-8 delivery along with documentation. 

Happy to discuss in an upcoming CalWG meeting. 

 

cc [Juergen Schreiber](https://jira.stsci.edu/secure/ViewProfile.jspa?name=schreibe) [Jeroen Bouwman](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bouwman) [Greg Sloan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=gcsloan) [Katherine Murray](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmurray)

### PR DESCRIPTION
**Pull request by hbushouse**
Closes #6277
Resolves [JP-2243](https://jira.stsci.edu/browse/JP-2243)

**Description**

This PR implements a pathloss correction for MIRI LRS fixed-slit exposures. It builds on the existing pathloss routines for NIRSpec slit-like modes. A new datamodel, `MirLrsPathlossModel`, is also implemented, because the form of the reference file data is different than those for NIRSpec or NIRISS SOSS (tabular instead of image-based). The `calwebb_spec2` pipeline is also updated to enable the pathloss step for LRS fixed-slit mode.

Note that a number of minor cosmetic changes have also been made to some of the existing pathloss functions for other modes, to clean up old errors in text and comments.

Checklist
- [x] Tests
- [x] Documentation
- [x] Change log
- [X] Milestone
- [X] Label(s)

### CODE REVIEW

File: docs/jwst/barshadow/arguments.rst [hunk 1]
```diff
@@ -1,5 +1,12 @@
 Step Arguments
```
Comment (hbushouse) [inline docs/jwst/barshadow/arguments.rst:1]: Note to reviewers: these updates to the barshadow, flatfield, and photom docs are not directly related to the LRS pathloss implementation, but I noticed while updated docs that these steps didn't list their available parameters.

**Comment by hbushouse** [comment]
@stscirij @tapastro @stscieisenhamer this is ready for review

File: jwst/pathloss/pathloss.py [hunk 2]
```diff
@@ -126,11 +148,15 @@ def calculate_pathloss_vector(pathloss_refdata,
     # pointsource.data is 3-d, so we have to extract a wavelength vector
     # at the specified location.  We do this using bilinear interpolation
     else:
-        crpix3 = pathloss_wcs.crpix3
-        crval3 = pathloss_wcs.crval3
-        cdelt3 = pathloss_wcs.cdelt3
-        for i in np.arange(wavesize):
-            wavelength[i] = crval3 + (float(i + 1) - crpix3) * cdelt3
+
+        # If requested, calculate a wavelength vector from the ref file
```
Comment (tapastro) [inline jwst/pathloss/pathloss.py:152]: Most minor of nitpicks, but perhaps change wording to indicate the default is on as well? "Unless not requested?" I'm not much of a wordsmith...

**Review by stscieisenhamer — APPROVED** [review]

**Review by tapastro — APPROVED** [review]
LGTM, with some small comments.

File: jwst/pathloss/pathloss.py [hunk 3]
```diff
@@ -854,3 +929,62 @@ def _corrections_for_ifu(data, pathloss, source_type):
...
+        pathloss_vector = pathloss_vector[::-1]
+
+        # Compute the point source pathloss 2D correction
+        pathloss_2d = interpolate_onto_grid(wavelength_array,
+                                            wavelength_vector,
+                                            pathloss_vector)
+
+        # Save the corrections. The `data` portion is the correction used.
+        # The individual ones will be saved in the respective attributes.
+        correction = type(data)(data=pathloss_2d)
```
[the remark this answers is not in the document]
Comment (hbushouse) [reply jwst/pathloss/pathloss.py:983]: I wasn't sure either when I first encountered it, but in this case `data` is the input data model, so `type(data)` is creating an object of the same type as `data` (for MIRI LRS it happens to be `ImageModel`). So it's a generic way of creating a new object that matches the input. But it does get confusing when you see the multiple occurrences of `data` in the statement ...

File: jwst/datamodels/schemas/mirlrs_pathloss.schema.yaml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+              fits_keyword: CUNIT1
+              fits_hdu: PATHLOSS
+            cunit2:
+              title: Axis 2 units
+              type: string
+              fits_keyword: CUNIT2
+              fits_hdu: PATHLOSS
+- type: object
+  properties:
+    meta:
```
[the remark this answers is not in the document]
Comment (hbushouse) [reply jwst/datamodels/schemas/mirlrs_pathloss.schema.yaml:97]: So this structure of the schema results in all of the "normal" meta attributes showing up within the `model.meta` tree, but the table column units attributes are within `model.meta.pathloss_table.tunit[123]`. Not sure we need it that way, but it works, and also makes it obvious that those meta attributes are strictly associated with the table part of the model.

**Pull request merged**